packages/main/src/TableSelection.ts

```diff
@@ -204,6 +216,180 @@ class TableSelection extends UI5Element implements ITableFeature {
+		}
+	}
+
+	_onclick(e: MouseEvent) {
+		if (!this._table) {
+			return;
+		}
+
+		if (this._isHeaderSelector(e)) {
+			this._informHeaderRowSelectionChange();
+			this._stopRangeSelection();
+			return;
+		}
+
+		if (!this._isSelectionCheckbox(e)) {
+			this._stopRangeSelection();
+			return;
+		}
+
+		const row = this._findRowInPath(e.composedPath());
+
+		if (e.shiftKey && this._rangeSelection?.isMouse) {
+			const startRow = this._rangeSelection.rows[0];
+			const startIndex = this._table?.rows.indexOf(startRow);
+			const endIndex = this._table?.rows.indexOf(row);
+
+			if (startIndex === -1 || endIndex === -1 || startIndex === undefined || endIndex === undefined
+				|| row.key === startRow.key || row.key === this._rangeSelection.rows[this._rangeSelection.rows.length - 1].key) {
+				return;
+			}
+
+			const change = endIndex - startIndex;
+			this._handleRangeSelection(row, change);
+		} else if (row) {
+			this.informSelectionChange(row);
+			this._startRangeSelection(row, true);
+		}
+	}
+
+	/**
+	 * Start the range selection and initialises the range selection state
+	 * @param row starting row
+	 * @private
+	 */
+	_startRangeSelection(row: TableRow, isMouse = false) {
+		const selected = this.isSelected(row);
+		if (isMouse && !selected) {
+			// Do not initiate range selection if the row is not selected
+			return;
+		}
+
+		this._rangeSelection = {
+			selected,
+			isUp: null,
+			rows: [row],
+			isMouse,
+		};
+	}
+
+	/**
+	 * Handles the range selection
+	 * @param targetRow row that is currently focused
+	 * @param change indicates direction
+	 * @private
+	 */
+	_handleRangeSelection(targetRow: TableRow, change: number) {
+		if (!this._rangeSelection) {
+			return;
+		}
+
+		const isUp = change > 0;
+		this._rangeSelection.isUp ??= isUp;
+
+		const shouldReverseSelection = isUp !== this._rangeSelection.isUp && !this._rangeSelection.isMouse;
+		let selectionChanged = shouldReverseSelection && this.isSelected(targetRow);
+
+		if (shouldReverseSelection) {
+			this._reverseRangeSelection();
+		} else {
+			const rowIndex = this._table!.rows.indexOf(targetRow);
+			const [startIndex, endIndex] = [rowIndex, rowIndex - change].sort((a, b) => a - b);
+
+			selectionChanged = this._table?.rows.slice(startIndex, endIndex + 1).reduce((changed, row) => {
+				const isRowNotInSelection = !this._rangeSelection?.rows.includes(row);
+				const isRowSelectionDifferent = this.isSelected(row) !== this._rangeSelection!.selected;
+
+				if (isRowNotInSelection) {
+					this._rangeSelection?.rows.push(row);
+				}
+
+				// Workaround required due to bug with mouse, where the checkbox stays unchecked, but the row is selected.
+				// Even invalidation does not change the state of the checkbox. Maybe an issue with lit's 'intelligent' rendering?
+				if (this._rangeSelection?.isMouse) {
+					row.shadowRoot?.querySelector("#selection-component")?.setAttribute("checked", "");
+				}
+
+				this._selectRow(row, this._rangeSelection!.selected);
```

Comment:
with mouse range selection do we fire the change event per row?

Reply:
Nope. Event is only fired once when doing mouse range selection.

Due to moving the space, enter handling back to Row and HeaderRow, had to add some additional logic to ensure this.